src/test/validation_chainstatemanager_tests.cpp

```diff
@@ -769,4 +770,44 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_completion_hash_mismatch, Sna
     }
 }
 
+/** Helper function to parse args into args_man and return the result of applying them to opts */
+util::Result<void> SetOptsFromArgs(ArgsManager& args_man, ChainstateManager::Options& opts,
+                                   const std::vector<const char*>& args)
+{
+    std::vector<const char*> argv = {"ignore"};
+    argv.insert(argv.end(), args.begin(), args.end());
+
+    std::string error{};
+    BOOST_REQUIRE(args_man.ParseParameters(argv.size(), argv.data(), error));
+    return node::ApplyArgsManOptions(args_man, opts);
+}
+
+BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
+{
+    kernel::Notifications notifications{};
+    const ChainstateManager::Options options{
+        .chainparams = ::Params(),
+        .datadir = {},
+        .notifications = notifications};
+
+    auto set_opts = [this, options](const std::vector<const char*>& args) {
+        auto opts{options}; // create a copy to keep the original clean and reusable
+        BOOST_REQUIRE(SetOptsFromArgs(*this->m_node.args, opts, args));
+        return opts;
+    };
+
+    // test -assumevalid
+    BOOST_CHECK(!set_opts({}).assumed_valid_block.has_value());
+    BOOST_CHECK(set_opts({"-assumevalid=0"}).assumed_valid_block.value().IsNull());
```

Comment:
Thanks for covering these scenarios!

I checked to see how this fails for the previous error:
> unknown location:0: fatal error: in "validation_chainstatemanager_tests/chainstatemanager_args": std::bad_optional_access: bad_optional_access

i.e. it doesn't get to the `IsNull` check since it can't call `value()` on a null optional, right?

I think we can work around that by comparing the dereferenced optional instead:
```suggestion
    BOOST_CHECK_EQUAL(*set_opts({"-assumevalid=0"}).assumed_valid_block, uint256::ZERO);
```

which gives this error for the previously invalid setup:
> src/test/validation_chainstatemanager_tests.cpp:801: error: in "validation_chainstatemanager_tests/chainstatemanager_args": check *set_opts({"-assumevalid=0"}).assumed_valid_block == uint256::ZERO has failed [000000016bcc2e20000000016bcc2e20000000016bcc29a03b66800104140f00 != 0000000000000000000000000000000000000000000000000000000000000000]

Which I find a lot more revealing.

Reply:
> Can you run this in valgrind?

Thanks for the context @maflcko! Valgrind is currently isn't really [compatible with ARM-based Macs](https://github.com/LouisBrunner/valgrind-macos/issues/56).

> Dereferencing a nullopt is UB 

And we can't use `BOOST_CHECK_EQUAL(set_opts({"-assumevalid=0"}).assumed_valid_block, std::make_optional(uint256::ZERO))` since there's no standard way to display optionals, right?

Ok, in that case just consider the test fix validated, it failed before the fix, passes now.

Reply:
I've noted it and will give it a try next time, thank you

Reply:
Thanks, I think that would produce better error messages on failure.

Reply:
Would be nice to have std::optional support in BOOST_CHECK_EQUAL but this PR has already had quite a bit of churn so I'm going to limit the scope and leave as is given that it's not super relevant.

Reply:
Split out to pr [`aec1c90` (#30618)](https://github.com/bitcoin/bitcoin/pull/30618/commits/aec1c901118311eec5dc24104a7439377c8fd6de)